File: 2617. Minimum Number of Visited Cells in a Grid/Solution.py
class Solution:
    def minimumVisitedCells(self, grid: List[List[int]]) -> int:
        q = [[0, 0]]
        n, m, steps = len(grid), len(grid[0]), 1
        if n == 1 and m == 1: return 1
        visited = [[False for i in range(m)] for j in range(n)]
        while q:
            l = len(q)
            for i in range(l):
                x, y = q.pop(0)
                for i in range(1, grid[x][y] + 1):
                    nx, ny = x + i, y + i
                    if (nx == n - 1 and y == m - 1) or (x == n - 1 and ny == m - 1): return steps + 1
                    if nx < n and not visited[nx][y] and grid[nx][y] > 0:
                        visited[nx][y] = True
                        q.append([nx, y])
                    if ny < m and not visited[x][ny] and grid[x][ny] > 0:
                        visited[x][ny] = True
                        q.append([x, ny])
            steps += 1
        return -1

File: 2617. Minimum Number of Visited Cells in a Grid/Solution.java
class Solution {
    public int minimumVisitedCells(int[][] grid) {
        Queue<int[]> q = new LinkedList<>();
        int n = grid.length, m = grid[0].length, steps = 1;
        if (n == 1 && m == 1) return 1;
        boolean[][] visited = new boolean[n][m];
        q.offer(new int[] { 0, 0 });
        while (!q.isEmpty()) {
            int l = q.size();
            for (int j = 0; j < l; j++) {
                int[] curr = q.poll();
                for (int i = 1; i <= grid[curr[0]][curr[1]]; i++) {
                    int nx = curr[0] + i, ny = curr[1] + i;
                    if ((nx == n - 1 && curr[1] == m - 1) || curr[0] == n - 1 && ny == m - 1) return steps + 1;
                    if (nx < n && !visited[nx][curr[1]] && grid[nx][curr[1]] > 0) {
                        visited[nx][curr[1]] = true;
                        q.offer(new int[] { nx, curr[1] });
                    }
                    if (ny < m && !visited[curr[0]][ny] && grid[curr[0]][ny] > 0) {
                        visited[curr[0]][ny] = true;
                        q.offer(new int[] { curr[0], ny });
                    }
                }
            }
            steps++;
        }
        return -1;
    }
}